nautobot/core/apps/__init__.py

```diff
@@ -0,0 +1,249 @@
+                raise TypeError("All groups defined in a tab must be an instance of NavMenuGroup")
+            self.groups = groups
+
+
+class NavMenuGroup(PermissionsMixin):
+    """
+    Ths class represents a navigation menu group. This is built up from a name and a weight value. The name is
+    the display text and the weight defines its position in the navbar.
+
+    Links are specified as Django reverse URL strings.
+    Items are each specified as a list of NavMenuItem instances.
+    """
+
+    permissions = []
+    items = []
+
+    def __init__(self, name, items=None, weight=1000):
+        """Ensure group properties."""
+        self.name = name
+        self.weight = weight
+
+        if items is not None and not isinstance(items, (list, tuple)):
+            raise TypeError("Items must be passed as a tuple or list.")
+        elif not all(isinstance(item, NavMenuItem) for item in items):
+            raise TypeError("All items defined in a group must be an instance of NavMenuItem")
+        self.items = items
+
+
+class NavMenuItem(PermissionsMixin):
+    """
+    This class represents a navigation menu item. This constitutes primary link and its text, but also allows for
+    specifying additional link buttons that appear to the right of the item in the nav menu.
+
+    Links are specified as Django reverse URL strings.
+    Buttons are each specified as a list of NavMenuButton instances.
+    """
+
+    def __init__(self, link, link_text, permissions=None, buttons=None, weight=1000):
+        """Ensure item properties."""
+        super().__init__(permissions)
+        reverse(link)
+        self.link = link
+        self.link_text = link_text
+        self.weight = weight
+
+        if buttons is not None and not isinstance(buttons, (list, tuple)):
+            raise TypeError("Buttons must be passed as a tuple or list.")
+        elif not all(isinstance(button, NavMenuButton) or issubclass(button, NavMenuButton) for button in buttons):
+            raise TypeError("All buttons defined in an item must be an instance or subclass of NavMenuButton")
+        self.buttons = buttons
+
+
+class NavMenuButton(PermissionsMixin):
+    """
+    This class represents a button within a PluginMenuItem. Note that button colors should come from
+    ButtonColorChoices.
+    """
+
+    def __init__(
+        self,
+        link,
+        title,
+        icon_class,
+        button_class=ButtonActionColorChoices.DEFAULT,
+        permissions=None,
+        weight=1000,
+    ):
+        """Ensure button properties."""
+        super().__init__(permissions)
+        reverse(link)
+        self.link = link
+        self.title = title
+        self.icon_class = icon_class
+        self.weight = weight
+
+        if button_class is not None and button_class not in ButtonActionColorChoices.values():
+            raise ValueError("Button color must be a choice within ButtonColorChoices.")
```

Comment:
This may be overly restrictive - for plugins in particular, we might want to allow the implementer to provide a non-standard class if none of the standard choices quite match their button's purpose/meaning.

Definitely if we go in the direction of making `icon_class` into a ChoiceSet option as I suggest in another comment, that should *not* be restricted to the ChoiceSet as plugin authors may wish to use any of the hundreds of other MDI icons as desired.